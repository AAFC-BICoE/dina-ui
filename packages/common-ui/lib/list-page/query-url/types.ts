export interface SimpleQueryGroup {
  // Conjunction (AND/OR)
  c: string;

  // Props: Items inside the group
  p: SimpleQueryRow[];
}

export interface SimpleQueryRow {
  // Field
  f: string;

  // Operator
  o: string;

  // Value
  v: string;

  // Type
  t: string;

  // Dynamic field identifier (UUID for example if managed attribute)
  d?: string;
}
