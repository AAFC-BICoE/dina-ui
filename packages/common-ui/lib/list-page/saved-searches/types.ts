import { JsonTree } from "react-awesome-query-builder";

/**
 * Since saved searches are stored as JSON this is the structure to be used.
 */
export interface SavedSearchStructure {
  [indexName: string]: {
    [savedSearchName: string]: SingleSavedSearch;
  };
}

export interface SingleSavedSearch {
  // This field is not saved in the JSON.
  savedSearchName?: string;
  default: boolean;
  queryTree?: JsonTree;
}
