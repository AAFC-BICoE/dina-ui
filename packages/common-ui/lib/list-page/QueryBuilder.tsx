import {
  DinaFormSection,
  FieldWrapperProps,
  LoadingSpinner,
  SelectField,
  useApiClient
} from "..";
import { QueryRow } from "./QueryRow";
import { FieldArray } from "formik";
import { GroupSelectField } from "../../../dina-ui/components";
import { useEffect, useMemo } from "react";
import { v4 as uuidv4 } from "uuid";
import useSWR from "swr";
import { ESIndexMapping } from "./types";

import {
  Query,
  Builder,
  Utils as QbUtils,
  BasicConfig
} from "react-awesome-query-builder";
import { useState, useCallback } from "react";
import {
  JsonGroup,
  Config,
  ImmutableTree,
  BuilderProps
} from "react-awesome-query-builder";
import { Button } from "react-bootstrap";
import { FaTrash } from "react-icons/fa";
import { QueryFieldSelector } from "./QueryFieldSelector";

interface QueryBuilderProps extends FieldWrapperProps {
  indexName: string;
  onGroupChange: (currentSubmittedValues: any) => void;
}

export function QueryBuilder({
  name,
  indexName,
  onGroupChange
}: QueryBuilderProps) {
  const { apiClient } = useApiClient();

  // State to store the index map after it has been retrieved.
  const [indexMap, setIndexMap] = useState<ESIndexMapping[]>([]);

  // Query Builder Configuration, only needs to update if the index map has changed.
  const config: Config = useMemo(
    () => ({
      conjunctions: {
        ...BasicConfig.conjunctions
      },
      operators: {
        equals: {
          label: "equals"
        }
      },
      widgets: {
        ...BasicConfig.widgets
      },
      types: {
        ...BasicConfig.types
      },
      fields: {
        test: {
          label: "test",
          type: "number"
        }
      },
      settings: {
        ...BasicConfig.settings,
        renderButton: (buttonProps) => {
          if (buttonProps) {
            switch (buttonProps?.type) {
              case "addRule":
              case "addGroup":
                return (
                  <Button onClick={buttonProps?.onClick} className="ms-1">
                    {buttonProps.label}
                  </Button>
                );
              case "delGroup":
              case "delRule":
              case "delRuleGroup":
                return (
                  <Button
                    onClick={buttonProps?.onClick}
                    className="ms-1"
                    variant="danger"
                  >
                    <FaTrash />
                  </Button>
                );
            }
          }

          return (
            <Button onClick={buttonProps?.onClick} className="ms-1">
              {buttonProps?.label}
            </Button>
          );
        },
        renderField: (fieldDropdownProps) => (
          <QueryFieldSelector indexMap={indexMap} />
        ),
        showNot: false,
        canRegroup: true,
        canReorder: true
      }
    }),
    [indexMap]
  );

  // State to store the query tree generated by the Query Builder.
  const [queryTree, setQueryTree] = useState<ImmutableTree>(
    QbUtils.checkTree(
      QbUtils.loadTree({
        id: QbUtils.uuid(),
        type: "group"
      }),
      config
    )
  );

  const onChange = useCallback((immutableTree: ImmutableTree) => {
    setQueryTree(immutableTree);
  }, []);

  const renderBuilder = useCallback(
    (props: BuilderProps) => (
      <div className="query-builder-container">
        <div className="query-builder qb-lite">
          <Builder {...props} />
        </div>
      </div>
    ),
    []
  );

  // Retrieve the index mapping.
  useEffect(() => {
    async function getIndexMapping() {
      const mapping = await fetchQueryFieldsByIndex(indexName);
      setIndexMap(mapping);
    }
    getIndexMapping();
  }, []);

  /**
   * The query builder options are generated from the elastic search index. This method will
   * request the mappings from the index.
   *
   * @param searchIndexName index to retrieve from.
   * @returns ESIndexMapping[]
   */
  async function fetchQueryFieldsByIndex(searchIndexName) {
    const resp = await apiClient.axios.get("search-api/search-ws/mapping", {
      params: { indexName: searchIndexName }
    });

    const result: ESIndexMapping[] = [];

    // Read index attributes.
    resp.data?.attributes
      ?.filter((key) => key.name !== "type")
      .map((key) => {
        const path = key.path;
        const prefix = "data.attributes";
        let attrPrefix;
        if (path && path.includes(prefix)) {
          attrPrefix = path.substring(prefix.length + 1);
        }
        result.push({
          label: attrPrefix ? attrPrefix + "." + key.name : key.name,
          value: key.path
            ? key.path + "." + key.name
            : key.name === "id"
            ? "data." + key.name
            : key.name,
          type: key.type,
          path: key.path,
          distinctTerm: key.distinct_term_agg
        });
      });

    // Read relationship attributes.
    resp.data?.relationships?.map((relationship) => {
      relationship?.attributes?.map((relationshipAttribute) => {
        // This is the user-friendly label to display on the search dropdown.
        const attributeLabel = relationshipAttribute.path?.includes(".")
          ? relationshipAttribute.path.substring(
              relationshipAttribute.path.indexOf(".") + 1
            ) +
            "." +
            relationshipAttribute.name
          : relationshipAttribute.name;

        result.push({
          label: attributeLabel,
          value: relationship.value + "." + attributeLabel,
          type: relationshipAttribute.type,
          path: relationshipAttribute.path,
          parentName: relationship.referencedBy,
          parentType: relationship.value,
          parentPath: relationship.path,
          distinctTerm: relationshipAttribute.distinct_term_agg
        });
      });
    });
    return result;
  }

  // Display loading spinner when performing request for the index.
  if (!indexMap) {
    return <LoadingSpinner loading={true} />;
  }

  const sortedData = indexMap
    ?.sort((a, b) => a.label.localeCompare(b.label))
    .filter((prop) => !prop.label.startsWith("group"));

  return (
    <>
      <Query
        {...config}
        value={queryTree}
        onChange={onChange}
        renderBuilder={renderBuilder}
      />
      <FieldArray name={name}>
        {(fieldArrayProps) => {
          const elements: [] = fieldArrayProps.form.values.queryRows;

          function addRow() {
            fieldArrayProps.push(
              <QueryRow
                name={fieldArrayProps.name}
                indexName={indexName}
                esIndexMapping={sortedData as any}
                index={elements?.length ?? 0}
                removeRow={removeRow}
                addRow={addRow}
              />
            );
            // initialize the logic switch value to be "and"
            fieldArrayProps.form.setFieldValue(
              `${fieldArrayProps.name}[${
                elements?.length ?? 0
              }].compoundQueryType`,
              "and"
            );
          }

          function removeRow(index) {
            fieldArrayProps.remove(index);
          }

          return elements?.length > 0
            ? elements?.map((_, index) => (
                <QueryRow
                  name={fieldArrayProps.name}
                  indexName={indexName}
                  key={index}
                  index={index}
                  addRow={addRow}
                  removeRow={removeRow}
                  esIndexMapping={sortedData as any}
                />
              ))
            : null;
        }}
      </FieldArray>
      <DinaFormSection horizontal={"flex"}>
        <GroupSelectField
          isMulti={true}
          name="group"
          className="col-md-4"
          onChange={(value, formik) =>
            onGroupChange({
              submittedValues: { ...formik.values, group: value }
            })
          }
        />
      </DinaFormSection>
    </>
  );
}
