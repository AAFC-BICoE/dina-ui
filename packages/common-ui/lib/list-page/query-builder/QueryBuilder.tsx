import { LoadingSpinner } from "../..";
import { Query, Builder, Utils, JsonTree } from "react-awesome-query-builder";
import { useCallback } from "react";
import {
  Config,
  ImmutableTree,
  BuilderProps
} from "react-awesome-query-builder";
import React from "react";
import { Button } from "react-bootstrap";
import { SavedSearch } from "../saved-searches/SavedSearch";
import { DinaMessage } from "../../../../dina-ui/intl/dina-ui-intl";

interface QueryBuilderProps {
  /**
   * Index name being used for the QueryPage.
   */
  indexName: string;

  /**
   * Query Builder Configuration.
   */
  queryBuilderConfig?: Config;

  /**
   * Query Builder local tree.
   */
  queryBuilderTree?: ImmutableTree;

  /**
   * Callback function to indicate that the query builder has changed.
   */
  setQueryBuilderTree: (newTree: ImmutableTree) => void;

  /**
   * Callback function to indicate the form was submitted.
   */
  onSubmit: () => void;

  /**
   * Callback to indicate the query builder was reset.
   */
  onReset: () => void;
}

function QueryBuilder({
  indexName,
  queryBuilderConfig,
  queryBuilderTree,
  setQueryBuilderTree,
  onSubmit,
  onReset
}: QueryBuilderProps) {
  const onChange = useCallback((immutableTree: ImmutableTree) => {
    setQueryBuilderTree(immutableTree);
  }, []);

  const renderBuilder = useCallback(
    (props: BuilderProps) => (
      <div className="query-builder-container">
        <div className="query-builder qb-lite">
          <Builder {...props} />
        </div>
      </div>
    ),
    []
  );

  // Display loading spinner when performing request for the index.
  if (!queryBuilderTree || !queryBuilderConfig) {
    return <LoadingSpinner loading={true} />;
  }

  return (
    <>
      <label
        style={{
          fontSize: 20,
          fontFamily: "sans-serif",
          fontWeight: "bold"
        }}
        className="mb-2"
      >
        <DinaMessage id="search" />
      </label>
      <SavedSearch
        indexName={indexName}
        queryBuilderTree={queryBuilderTree}
        setQueryBuilderTree={setQueryBuilderTree}
        queryBuilderConfig={queryBuilderConfig}
        performSubmit={onSubmit}
      />
      <Query
        {...queryBuilderConfig}
        value={queryBuilderTree}
        onChange={onChange}
        renderBuilder={renderBuilder}
      />
      <div className="mt-2">
        <Button onClick={onSubmit} className="me-2">
          Search
        </Button>
        <Button onClick={onReset} variant="secondary">
          Reset
        </Button>
      </div>
    </>
  );
}

/**
 * Default query tree contains an empty rule so it's one less the step the user needs to perform.
 *
 * This is the default when first loading the page and resetting the query builder.
 */
export function defaultQueryTree(): ImmutableTree {
  const groupId = "8c6dc2c8-4070-48ce-b700-13a931f9ebaf";
  const ruleId = "f76a54f6-0112-4ac9-b2a1-f6dced58b3d6";

  return Utils.loadTree({
    id: groupId,
    type: "group",
    children1: {
      "f76a54f6-0112-4ac9-b2a1-f6dced58b3d6": {
        type: "rule",
        id: ruleId,
        properties: {
          field: null,
          operator: null,
          value: [],
          valueSrc: [],
          valueError: []
        },
        path: [groupId, ruleId]
      }
    },
    properties: { conjunction: "AND" },
    path: [groupId]
  } as JsonTree);
}

/**
 * Generate an empty tree with no rules. Used for special cases like custom elastic search
 * view queries where the query builder is not needed.
 */
export function emptyQueryTree(): ImmutableTree {
  return Utils.loadTree({
    id: Utils.uuid(),
    type: "group",
    children1: {}
  } as JsonTree);
}

/**
 * Currently utilized for the custom view queries to show associated records.
 *
 * @param uuid UUID to search for.
 * @param path Relationship path to search against.
 * @returns QueryBuilder tree which will be translated into elastic search query.
 */
export function generateUUIDTree(uuid: string, path: string): JsonTree {
  return {
    id: Utils.uuid(),
    type: "group",
    children1: {
      [Utils.uuid()]: {
        type: "rule",
        properties: {
          field: path,
          operator: "uuid",
          value: [uuid]
        }
      }
    }
  };
}

export const QueryBuilderMemo = React.memo(QueryBuilder);
