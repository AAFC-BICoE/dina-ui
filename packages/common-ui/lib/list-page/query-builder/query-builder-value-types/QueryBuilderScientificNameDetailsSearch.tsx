import React, { useMemo, useState } from "react";
import { useIntl } from "react-intl";
import Select from "react-select";
import { useEffect } from "react";
import { noop, startCase } from "lodash";
import { useQueryBuilderEnterToSearch } from "../query-builder-core-components/useQueryBuilderEnterToSearch";
import { VocabularyOption } from "../../../../../dina-ui/components/collection/VocabularySelectField";
import useVocabularyOptions from "../../../../../dina-ui/components/collection/useVocabularyOptions";
import { SelectOption } from "packages/common-ui/lib/formik-connected/SelectField";

interface QueryRowScientificNameDetailsSearchProps {
  /**
   * Retrieve the current value from the Query Builder.
   */
  value?: string;

  /**
   * Pass the selected value to the Query Builder to store.
   */
  setValue?: (fieldPath: string) => void;

  /**
   * If being used in the column selector, operators do not need to be displayed.
   */
  isInColumnSelector: boolean;
}

export interface ScientificNameDetailsSearchStates {
  selectedClassificationRank: string;

  /** Operator to be used on the scientific name details search. */
  selectedOperator: string;

  /** The value the user wishes to search. */
  searchValue: string;
}

export default function QueryRowScientificNameDetailsSearch({
  value,
  setValue,
  isInColumnSelector
}: QueryRowScientificNameDetailsSearchProps) {
  const { formatMessage } = useIntl();

  // Used for submitting the query builder if pressing enter on a text field inside of the QueryBuilder.
  const onKeyDown = isInColumnSelector ? noop : useQueryBuilderEnterToSearch();

  const [scientificNameDetailsState, setScientificNameDetailsState] =
    useState<ScientificNameDetailsSearchStates>(() =>
      value
        ? JSON.parse(value)
        : {
            searchValue: "",
            selectedOperator: "",
            selectedClassificationRank: ""
          }
    );

  // Convert the state in this component to a value that can be stored in the Query Builder.
  useEffect(() => {
    if (setValue) {
      setValue(JSON.stringify(scientificNameDetailsState));
    }
  }, [scientificNameDetailsState]);

  // Convert a value from Query Builder into the Scientific Name Details State in this component.
  // Dependent on the identifierConfig to indicate when it's changed.
  useEffect(() => {
    if (value) {
      setScientificNameDetailsState(JSON.parse(value));
    }
  }, []);

  // Supported operators for the classification search.
  const supportedOperators: string[] = [
    "exactMatch",
    "wildcard",
    "in",
    "notIn",
    "startsWith",
    "notEquals",
    "empty",
    "notEmpty"
  ];

  // Retrieve the classification options
  const { loading, vocabOptions: taxonomicRankOptions } = useVocabularyOptions({
    path: "collection-api/vocabulary2/taxonomicRank"
  });

  // Capitalize each label for the taxonomic rank options.
  useMemo(() => {
    taxonomicRankOptions.forEach((option) => {
      option.label = startCase(option.label);
    });
  }, [taxonomicRankOptions]);

  // Generate the operator options
  const operatorOptions = supportedOperators.map<SelectOption<string>>(
    (option) => ({
      label: formatMessage({ id: "queryBuilder_operator_" + option }),
      value: option
    })
  );

  // Currently selected option, if no option can be found just select the first one.
  const selectedOperator = operatorOptions?.find(
    (operator) => operator.value === scientificNameDetailsState.selectedOperator
  );

  return (
    <div className={isInColumnSelector ? "" : "row"}>
      {/* Classification Rank Selection */}
      <Select<VocabularyOption>
        options={taxonomicRankOptions}
        value={taxonomicRankOptions.find(
          (option) =>
            option.value ===
            scientificNameDetailsState.selectedClassificationRank
        )}
        isLoading={loading}
        placeholder={formatMessage({
          id: "queryBuilder_scientificNameDetails_placeholder"
        })}
        onChange={(newValue) => {
          setScientificNameDetailsState({
            ...scientificNameDetailsState,
            selectedClassificationRank: newValue?.value ?? "",
            selectedOperator: "",
            searchValue: ""
          });
        }}
        controlShouldRenderValue={true}
        isClearable={false}
        className={isInColumnSelector ? "ps-0 mt-2 mb-3" : "col me-1 ms-2 ps-0"}
        onKeyDown={onKeyDown}
        captureMenuScroll={true}
        menuPlacement={isInColumnSelector ? "bottom" : "auto"}
        menuShouldScrollIntoView={false}
        minMenuHeight={600}
      />

      {/* Operator Selection */}
      {!isInColumnSelector &&
        scientificNameDetailsState.selectedClassificationRank !== "" && (
          <Select<SelectOption<string>>
            options={operatorOptions}
            className={`col me-1 ps-0`}
            value={selectedOperator}
            onChange={(selected) =>
              setScientificNameDetailsState({
                ...scientificNameDetailsState,
                selectedOperator: selected?.value ?? ""
              })
            }
            captureMenuScroll={true}
            menuPlacement={"auto"}
            menuShouldScrollIntoView={false}
            minMenuHeight={600}
          />
        )}

      {/* Search Value Input */}
      {!isInColumnSelector &&
        scientificNameDetailsState.selectedClassificationRank !== "" &&
        scientificNameDetailsState.selectedOperator !== "empty" &&
        scientificNameDetailsState.selectedOperator !== "notEmpty" && (
          <input
            type="text"
            value={scientificNameDetailsState.searchValue}
            onChange={(newValue) => {
              setScientificNameDetailsState({
                ...scientificNameDetailsState,
                searchValue: newValue.target.value
              });
            }}
            className={"col form-control me-3"}
            placeholder={
              scientificNameDetailsState.selectedOperator !== "in" &&
              scientificNameDetailsState.selectedOperator !== "notIn"
                ? formatMessage({
                    id: "queryBuilder_value_text_placeholder"
                  })
                : formatMessage({ id: "queryBuilder_value_in_placeholder" })
            }
            onKeyDown={onKeyDown}
          />
        )}
    </div>
  );
}

/**
 * Using the query row for a target organism primary classification search, generate the elastic
 * search request to be made.
 */
// export function transformClassificationToDSL({
//   value,
//   fieldInfo,
//   indexMap
// }: TransformToDSLProps): any {

// }
