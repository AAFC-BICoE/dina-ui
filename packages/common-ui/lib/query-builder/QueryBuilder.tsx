import { useMemo } from "react";
import useSWR from "swr";
import { DinaForm, useApiClient } from "..";
import { QueryRow } from "./QueryRow";
import { v4 as uuidv4 } from "uuid";

export function QueryBuilder() {
  const { apiClient } = useApiClient();

  async function fetchQueryFieldsByIndex(indexName) {
    const resp = await apiClient.axios.get("search-api/search/mapping", {
      params: { indexName }
    });

    const result: [{}] = [{}];

    Object.keys(resp.data)
      .filter(key => key.includes("data.attributes."))
      .map(key => {
        const fieldNameLabel = key.substring(
          "data.attributes.".length,
          key.lastIndexOf(".")
        );
        result.push({
          label: fieldNameLabel,
          value: fieldNameLabel,
          type: resp.data?.[key]
        });
      });

    return result;
  }

  // Invalidate the query cache on query change, don't use SWR's built-in cache:
  const cacheId = useMemo(() => uuidv4(), []);

  const {
    data,
    error,
    isValidating: loading
  } = useSWR(["dina_material_sample_index", cacheId], fetchQueryFieldsByIndex, {
    shouldRetryOnError: true,
    revalidateOnFocus: false,
    revalidateOnReconnect: false
  });

  if (loading || error) return <></>;

  return (
    <DinaForm initialValues={{}}>
      <QueryRow queryRowProps={data} />
    </DinaForm>
  );
}
