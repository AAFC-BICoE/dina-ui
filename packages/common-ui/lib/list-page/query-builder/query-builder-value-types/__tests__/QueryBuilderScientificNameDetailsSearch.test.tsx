import QueryBuilderScientificNameDetailsSearch from "../QueryBuilderScientificNameDetailsSearch";
import { mountWithAppContext } from "common-ui/lib/test-util/mock-app-context";
import { DinaForm } from "common-ui/lib/formik-connected/DinaForm";
import { QueryBuilderContextProvider } from "../../QueryBuilder";
import { noop } from "lodash";
import userEvent from "@testing-library/user-event";
import { Vocabulary } from "packages/dina-ui/types/collection-api";

export const TEST_CLASSIFICATIONS: Vocabulary = {
  id: "taxonomicRank",
  type: "vocabulary",
  vocabularyElements: [
    {
      key: "kingdom",
      name: "kingdom",
      multilingualTitle: {
        titles: [
          {
            lang: "en",
            title: "kingdom"
          },
          {
            lang: "fr",
            title: "r√®gne"
          }
        ]
      }
    }
  ]
};

const mockGet = jest.fn<any, any>(async (path) => {
  switch (path) {
    case "collection-api/vocabulary2/taxonomicRank":
      return { data: TEST_CLASSIFICATIONS };
  }
});

const apiContext: any = {
  apiClient: { get: mockGet }
};

describe("QueryBuilderScientificNameDetailsSearch", () => {
  describe("QueryBuilderScientificNameDetailsSearch Component", () => {
    it("Display field if match type is equals", async () => {
      // This test will just ensure the layout does not change unexpectedly.
      // Any changes to the layout, the snapshots will need to be updated.
      const textSearchEquals = mountWithAppContext(
        <DinaForm initialValues={{}}>
          <QueryBuilderContextProvider
            value={{ performSubmit: noop, groups: [] }}
          >
            <QueryBuilderScientificNameDetailsSearch
              value='{"searchValue":"","selectedOperator":"exactMatch","selectedClassificationRank":"class"}'
              setValue={jest.fn}
              isInColumnSelector={false}
            />
          </QueryBuilderContextProvider>
        </DinaForm>,
        {
          apiContext
        }
      );

      // Expect a snapshot with the text field being displayed.
      expect(textSearchEquals.asFragment()).toMatchSnapshot(
        "Expect text field to be displayed since match type is equals"
      );

      const textSearchEmpty = mountWithAppContext(
        <DinaForm initialValues={{}}>
          <QueryBuilderContextProvider
            value={{ performSubmit: noop, groups: [] }}
          >
            <QueryBuilderScientificNameDetailsSearch
              value='{"searchValue":"","selectedOperator":"empty","selectedClassificationRank":"class"}'
              setValue={jest.fn}
              isInColumnSelector={false}
            />
          </QueryBuilderContextProvider>
        </DinaForm>,
        {
          apiContext
        }
      );

      // Expect a snapshot without the text field being displayed.
      expect(textSearchEmpty.asFragment()).toMatchSnapshot(
        "Expect text field not to be displayed since the match type is not equals"
      );
    });

    it("Should call performSubmit on enter key press in textfield", async () => {
      const mockPerformSubmit = jest.fn();
      const wrapper = mountWithAppContext(
        <DinaForm initialValues={{}}>
          <QueryBuilderContextProvider
            value={{ performSubmit: mockPerformSubmit, groups: [] }}
          >
            <QueryBuilderScientificNameDetailsSearch
              value='{"searchValue":"","selectedOperator":"exactMatch","selectedClassificationRank":"class"}'
              setValue={jest.fn}
              isInColumnSelector={false}
            />
          </QueryBuilderContextProvider>
        </DinaForm>,
        {
          apiContext
        }
      );

      // Find the text field element
      const textField = wrapper.getByRole("textbox");

      // Expect performSubmit to not be called yet.
      expect(mockPerformSubmit).toHaveBeenCalledTimes(0);

      // Simulate user typing "enter" key
      userEvent.type(textField, "{enter}");

      // Expect performSubmit to be called once
      expect(mockPerformSubmit).toHaveBeenCalledTimes(1);
    });
  });
});
