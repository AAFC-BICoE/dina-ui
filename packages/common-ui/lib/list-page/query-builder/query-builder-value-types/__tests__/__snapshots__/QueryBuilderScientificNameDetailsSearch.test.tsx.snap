// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryBuilderScientificNameDetailsSearch QueryBuilderScientificNameDetailsSearch Component Display field if match type is equals: Expect text field not to be displayed since the match type is not equals 1`] = `
<DocumentFragment>
  <div>
    <form
      action="#"
    >
      <div
        class="error-viewer"
      />
      <div
        class="row"
      >
        <div
          class="col me-1 ms-2 ps-0 css-b62m3t-container"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-4-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
            role="log"
          />
          <div
            class=" css-13cymwt-control"
          >
            <div
              class=" css-1fdsijx-ValueContainer"
            >
              <div
                class=" css-1jqq78o-placeholder"
                id="react-select-4-placeholder"
              >
                Select classification rank...
              </div>
              <div
                class=" css-qbdosj-Input"
                data-value=""
              >
                <input
                  aria-activedescendant=""
                  aria-autocomplete="list"
                  aria-describedby="react-select-4-placeholder"
                  aria-expanded="false"
                  aria-haspopup="true"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class=""
                  id="react-select-4-input"
                  role="combobox"
                  spellcheck="false"
                  style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class=" css-1hb7zxy-IndicatorsContainer"
            >
              <div
                aria-hidden="true"
                class=" css-pogzpp-loadingIndicator"
              >
                <span
                  class="css-1xtdfmb-LoadingDot"
                />
                <span
                  class="css-zoievk-LoadingDot"
                />
                <span
                  class="css-x748d8-LoadingDot"
                />
              </div>
              <span
                class=" css-1u9des2-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class=" css-1xc3v61-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <div
          class="col me-1 ps-0 css-b62m3t-container"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-5-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
            role="log"
          />
          <div
            class=" css-13cymwt-control"
          >
            <div
              class=" css-1fdsijx-ValueContainer"
            >
              <div
                class=" css-1dimb5e-singleValue"
              >
                Empty
              </div>
              <div
                class=" css-qbdosj-Input"
                data-value=""
              >
                <input
                  aria-activedescendant=""
                  aria-autocomplete="list"
                  aria-expanded="false"
                  aria-haspopup="true"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class=""
                  id="react-select-5-input"
                  role="combobox"
                  spellcheck="false"
                  style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class=" css-1hb7zxy-IndicatorsContainer"
            >
              <span
                class=" css-1u9des2-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class=" css-1xc3v61-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`QueryBuilderScientificNameDetailsSearch QueryBuilderScientificNameDetailsSearch Component Display field if match type is equals: Expect text field to be displayed since match type is equals 1`] = `
<DocumentFragment>
  <div>
    <form
      action="#"
    >
      <div
        class="error-viewer"
      />
      <div
        class="row"
      >
        <div
          class="col me-1 ms-2 ps-0 css-b62m3t-container"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-2-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
            role="log"
          />
          <div
            class=" css-13cymwt-control"
          >
            <div
              class=" css-1fdsijx-ValueContainer"
            >
              <div
                class=" css-1jqq78o-placeholder"
                id="react-select-2-placeholder"
              >
                Select classification rank...
              </div>
              <div
                class=" css-qbdosj-Input"
                data-value=""
              >
                <input
                  aria-activedescendant=""
                  aria-autocomplete="list"
                  aria-describedby="react-select-2-placeholder"
                  aria-expanded="false"
                  aria-haspopup="true"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class=""
                  id="react-select-2-input"
                  role="combobox"
                  spellcheck="false"
                  style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class=" css-1hb7zxy-IndicatorsContainer"
            >
              <div
                aria-hidden="true"
                class=" css-pogzpp-loadingIndicator"
              >
                <span
                  class="css-1xtdfmb-LoadingDot"
                />
                <span
                  class="css-zoievk-LoadingDot"
                />
                <span
                  class="css-x748d8-LoadingDot"
                />
              </div>
              <span
                class=" css-1u9des2-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class=" css-1xc3v61-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <div
          class="col me-1 ps-0 css-b62m3t-container"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-3-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
            role="log"
          />
          <div
            class=" css-13cymwt-control"
          >
            <div
              class=" css-1fdsijx-ValueContainer"
            >
              <div
                class=" css-1dimb5e-singleValue"
              >
                Exact match
              </div>
              <div
                class=" css-qbdosj-Input"
                data-value=""
              >
                <input
                  aria-activedescendant=""
                  aria-autocomplete="list"
                  aria-expanded="false"
                  aria-haspopup="true"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class=""
                  id="react-select-3-input"
                  role="combobox"
                  spellcheck="false"
                  style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class=" css-1hb7zxy-IndicatorsContainer"
            >
              <span
                class=" css-1u9des2-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class=" css-1xc3v61-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <input
          class="col form-control me-3"
          placeholder="Enter text search value..."
          type="text"
          value=""
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;
